### ISSUE TITLE
Running the Pool

### ISSUE THREAD

**Issue by stone212** [sammy007/open-ethereum-pool#234]
#### System information

Geth version: pARITY 1.7.9

Redis version: 3.0.6

OS & Version: Ubuntu 16.04

Commit branch and hash: master

#### Expected behaviour

@sammy007 has said that payouts will "just work"?

#### Actual behaviour

Payouts don't work.  All of the mined coin goes to the "author".  This is the output of running the pool:

```
2017/11/25 01:21:47 Loading config: /home/openethereumpool/open-ethereum-pool/payouts.json
2017/11/25 01:21:47 Running with 2 threads
2017/11/25 01:21:47 Backend check reply: PONG
2017/11/25 01:21:47 Starting payouts
2017/11/25 01:21:47 Set payouts interval to 10m0s
2017/11/25 01:21:47 No payees that have reached payout threshold
```

#### Steps to reproduce the behaviour

There was no documentation on Payouts, but in another thread @weldstech said

> payouts in referance to the payouts within the json file. within the json files are the modules hence the true or false to turn the on or off. to run payouts all other modules within that instance of the json file your about to run should be turned off. Example create another config.json but call it payments.json and only enable the payments module within it and then run that as an instance.

This was extremely helpful.  I created a new file and this is my payouts section:
```

       "payouts": {
                "enabled": true,
                "requirePeers": 1,
                "interval": "10m",
                "daemon": "http://127.0.0.1:8545",
                "timeout": "10s",
                "address": "0x00abbC34d5506C4397B3B920Ea9F2Bb210104799",
                "gas": "21000",
                "gasPrice": "50000000000",
                "autoGas": true,
                "threshold": 50,
                "bgsave": false
        },
```

The payee address is unlocked.

#### Backtrace

````
[backtrace]
````

**Comment by sammy007** [comment]
I am about to stop commenting on issues here or probably disable them at all, but well, what word in "No payees that have reached payout threshold" you don't understand?

**Comment by kulisu** [comment]
> No payees that have reached payout threshold

nooob ! your pool mined nothing so no rewards to miners.
you will figure out that payout module works as expected once you mined a block.

to Sammy, we're really appreciated your excellent project, thanks !
TWETH.tw

**Comment by stone212** [comment]
@sammy007 The author account has collected over 100,000 "eth" (mined by the payee on this private blockchain) so not reaching the threshold is not the problem.  

@kulisu I have mined over 25,000 blocks (on the private blockchain I am using the pool on).

**Comment by kulisu** [comment]
```bash
john_doe@tweth.tw:~$ redis-cli
127.0.0.1:6379> AUTH "foobared"
OK
127.0.0.1:6379> KEYS "*miners*"
   1) "eth:miners:0x0000000000000000000000000000000000000000"
   2) "eth:miners:0x1111111111111111111111111111111111111111"
<snipped>
1626) "eth:miners:0x2222222222222222222222222222222222222222"
1627) "eth:miners:0x3333333333333333333333333333333333333333"
1628) "eth:miners:0x4444444444444444444444444444444444444444"
1629) "eth:miners:0x5555555555555555555555555555555555555555"
127.0.0.1:6379> HGETALL "eth:miners:0x5555555555555555555555555555555555555555"
1) "lastShare"
2) "1510154510"
3) "immature"
4) "0"
5) "balance"
6) "5348661"
```

is there any miner record in your Redis ?
if answer is yes, miner's `balance` field is greater than zero ?

**Comment by stone212** [comment]
EDIT:

@kulisu This is very helpful.

Redis does show the two miners.

HGETALL for one of the miners returns:

```
1) "lastShare"
2) "1511598239"
3) "blocksFound"
4) "6987"
```

I recently re initialized the blockchain so 4 is lower than it was before.  Why does my readout look so different from yous?  No immature, 0, balance, etc.

**Comment by kulisu** [comment]
```bash
john_doe@tweth.tw:~$ redis-cli
127.0.0.1:6379> AUTH "foobared"
OK

127.0.0.1:6379> KEYS "*block*"
1) "eth:blocks:matured"

127.0.0.1:6379> ZRANGE "eth:blocks:matured" 0 -1 WITHSCORES
  1) "0:0:0xd5927e201333dad2:0xa3262db4d0516b35e25c17305f5c7776307ab9012dafecdebb8a22b2cf12760a:1502446540:1645301224453789:19487000000000:5000000000000000000"
  2) "4143977"
  3) "0:0:0x222df670175f6748:0x7b40a5d3a0ca63d4757aa0fa7da83d2782633bcaae813c4efe3c21cbcf1be9c3:1502504323:1656196565249555:2850740000000000:5000000000000000000"
<snipped>
685) "0:0:0xfb6af26012c987ee:0x340a893921c334db30143cd140ac26a6db2c5163bc5a5e969c7289a162b5e630:1511521063:1532349557062439:704817450000000:3148934389478608155"
686) "4612702"
687) "0:0:0x1474b87b46d9e8d5:0x38fd39e0c9cc9a54b2ca1c12a704a1280cffefb81c23b402e45525cc3f9fc692:1511576266:1553489862544969:1974973950000000:3033483078210472886"
688) "4616702"
689) "0:0:0x56671ba400656d28:0x263285c42c058d6bd6d3bb48abd4ee2161edde1be8c410a86799c24d1e214754:1511597149:1546887408828041:748495800000000:3041948453894770597"
690) "4618186"
691) "4618852:0:0x391bfe80047729e6:0x5f3263a85abaeebd917b526057141e0fb150e6e07cd194b1dc92fee60ec81c34:1511606115:1564863965038604:312695400000000:2250000000000000000"
692) "4618854"
```

please paste above Redis output here, and a quick question is ..
have you unlocked all blocks you mined ? and paste your unlock log here.

1. block nonce submitted > new blocks mined > unlocker module invoked
2. miners' shares calculated > miners' profit calculated > payout module invoked
3. payout threshold reached > pending `balance` sent > miners get profit

**Comment by stone212** [comment]
@kulisu 

```
127.0.0.1:6379> KEYS "*block*"
1) "eth:blocks:candidates"
```
So that is something right there.  Do you have any idea what it means?  I'm going to start searching right now.  But it looks like you led me to the problem (or at least one problem).

Then I get a whole list like yours when I type:

`127.0.0.1:6379> ZRANGE "eth:blocks:candidates" 0 -1 WITHSCORES
`

I also have a question.  You asked if I have "unlocked all the blocks" and I'm not sure if that's what the difference between matured and candidate is?  If so then can you tell me how to unlock them?  "Unlock" is something that I connect with accounts not blocks.  And all of the searches I did talk about unlocking accounts.

**Comment by kulisu** [comment]
OK, we can now realize the problem is you didnt **unlock the candidate blocks to let them go matured**.

### Unlocker Module

1. new blocks mined > block info written in `eth:blocks:candidates`.
2. unlocker module invoked > `eth:blocks:candidates` deleted > detail written in `eth:blocks:matured`.
3. miners' shares/profit calculated > `eth:miners:0xXXXX` _balance_ increased automatically.
4. `eth:finances` _totalMined_ calculated automatically.

### Payout Module

1. payout module invoked > check `eth:miners:0xXXXX` _balance_ greater than threshold.
2. pending balance sent > `eth:miners:0xXXXX` _balance_ zeroed > `eth:miners:0xXXXX` _paid_ increased.
3. `eth:finances` _paid_ increased > `eth:finances` _balance_ calculated automatically and so on.

please review your config.json file, ensure that unlocker module is enabled.
re-launch open-ethereum-pool binary and everything should works as expected.

```json
  "unlocker": {
    "enabled": true,
```

and the last IMPORTANT thing is READ THIS POOL CODE and think about what it works.
building a pool with zero knowledge is really dangerous, you will lose block rewards and miners, etc.

**this is a well-known project and be used by a bunch of ETC/ETH/EXP/Music/SILO/UBQ/Whale pools .**
**answering this question is really tired and bored for someone else (and worst you think this is a issue).**

good luck :smile:

**Comment by stone212** [comment]
> you didnt unlock the candidate blocks to let them go matured

Yes, this is most likely the problem.  Thank you for your help.  I see now how it is in the comments of the sample config file.  I was searching for keywords and "remit" was not one of the words.  

Now I just enabled "unlocker" and now I am getting a long list of Orphaned block/Mature block.  We will see what happens soon.

> READ THIS POOL CODE

You are very correct that I should have re-read the config.json more completely.  I read it completely months ago when I started but that time did not have the problem I have now.  I should have re-read it completely.  

> and worst you think this is a issue

Well, if there is not adequate documentation, there is an issue.  Issue does not mean bug.  I never said there is a bug.  But I was wrong that this is not documented as I just said so thank you.  And in the process of this Issue you gave me some very useful debugging information.  Thank you again.

This is still syncing so I will wait to close.  But it is definitely progress.
